### PR TITLE
[TD] Add CommandHoleShaftFit a new command creating hole / shaft fit using ISO 286

### PR DESCRIPTION
**Pull request by edi271**
New command creating hole /shaft fit using ISO 286 in TechDraw. This command does the same as the macro HoleShaftFitISO286.FCMacro which was presented in https://forum.freecad.org/viewtopic.php?t=71650&sid=9a92a404afff23b0c9a45d9974f6c6f4
Wiki will be updated in the next days.

Thank you for creating a pull request to contribute to FreeCAD! Place an "X" in between the brackets below to "check off" to confirm that you have satisfied the requirement, or ask for help in the [FreeCAD forum](https://forum.freecadweb.org/viewforum.php?f=10) if there is something you don't understand.

- [X]  Your Pull Request meets the requirements outlined in section 5 of [CONTRIBUTING.md](https://github.com/FreeCAD/FreeCAD/blob/master/CONTRIBUTING.md) for a Valid PR

Please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [1.0 Changelog Forum Thread](https://forum.freecad.org/viewtopic.php?f=10&t=69438).

---

### CODE REVIEW

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8455">pipeline status</a> for feature branch [PR_8455](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_8455). Pipeline [775210067](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/775210067) was triggered at [41c26a3](https://github.com/FreeCAD/FreeCAD/pull/8455/commits/41c26a3a521c9bded0171ed1dfd11b9fad89acf3). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

File: src/Mod/TechDraw/TechDrawTools/CommandHoleShaftFit.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        """Return True when the command should be active or False when it should be disabled (greyed)."""
+        if App.ActiveDocument:
+            return TechDrawTools.TDToolsUtil.havePage() and TechDrawTools.TDToolsUtil.haveView()
+        else:
+            return False
+
+#
+# The command must be "registered" with a unique name by calling its class.
+Gui.addCommand('TechDraw_HoleShaftFit', CommandHoleShaftFit())
+
```
Comment (luzpaz) [inline src/Mod/TechDraw/TechDrawTools/CommandHoleShaftFit.py:72]: Remove trailing newline please

File: src/Mod/TechDraw/TechDrawTools/CommandHoleShaftFit.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+import TechDrawTools
+
+class CommandHoleShaftFit:
+    """Adds a hole or shaft fit to a selected dimension."""
+
+    def __init__(self):
+        """Initialize variables for the command that must exist at all times."""
+        pass
+
```
Comment (benj5378) [inline src/Mod/TechDraw/TechDrawTools/CommandHoleShaftFit.py:42]:
```suggestion
```

File: src/Mod/TechDraw/TechDrawTools/TaskHoleShaftFit.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+from functools import partial
+
+import os
+
+class TaskHoleShaftFit:
+    def __init__(self,sel):
+
+        self.isHole = True
+        self.sel = sel
+        self.holeValues = [["h9","D10","loose"],["h9","E9","loose"],["h9","F8","Loose"],["h6","G7","loose"],
```
Comment (benj5378) [inline src/Mod/TechDraw/TechDrawTools/TaskHoleShaftFit.py:45]: `["h9","D10","loose"]` What does each of these values mean? I don't know anything about ISO holes... ELI5

File: src/Mod/TechDraw/TechDrawTools/TaskHoleShaftFit.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+
+__title__ = "TechDrawTools.TaskHoleShaftFit"
+__author__ = "edi"
+__url__ = "https://www.freecadweb.org"
+__version__ = "00.01"
+__date__ = "2023/02/07"
+
+from PySide.QtCore import QT_TRANSLATE_NOOP
+from PySide import QtCore
+import PySide.QtGui as QtGui
```
Comment (benj5378) [inline src/Mod/TechDraw/TechDrawTools/TaskHoleShaftFit.py:31]: Unused imports?

**Review by benj5378 — CHANGES_REQUESTED** [review]
Please run `black` and `flake8` on the python files

File: src/Mod/TechDraw/TechDrawTools/TaskHoleShaftFit.py (hunk 3 above)
Comment (WandererFan) [reply src/Mod/TechDraw/TechDrawTools/TaskHoleShaftFit.py:45]: https://www.cobanengineering.com/Tolerances/ISOShaftTolerances3mmTo400mm.asp

First column in the chart is the source of the "magic numbers".

Comment (edi271) [reply src/Mod/TechDraw/TechDrawTools/TaskHoleShaftFit.py:45]: Nothing can be produced exact. Therefore ISO286 defines quality fields. Assume you have to produce several shafts. The shafts do not need to have the exact diameter, their diameter must be inside a defined field.
For the following example see table posted above by Wandererfan:
Assume the specified size of the shafts diameter is 35mm. If our field is f7, all diameters must be less then the specified size. They have to be between minimum 25 micrometers, maximum 50 micrometers less. Maybe our shafts are the axes of rotating parts.
Lower case fields define shafts, upper case fields define holes. The bigger the number, the bigger the field range, compare h10, h11, h12. Letters a to g define undersize, j to z define oversize. The h and H fields are special fields, one value is 0.
There are always two parts combined, one shaft and one hole. Three types of fit are possible: loose fit, snug fit and press fit.
["h9","D10","loose"] means:
Shaft has field h quality 9. Hole has field D, quality 10. The type of fit is loose, a moving between the parts is possible. In all combinations one of the two parts has the h (or H) field.

File: src/Mod/TechDraw/TechDrawTools/CommandHoleShaftFit.py [hunk 5]
```diff
@@ -68,5 +64,4 @@ def IsActive(self):
 
 #
 # The command must be "registered" with a unique name by calling its class.
-Gui.addCommand('TechDraw_HoleShaftFit', CommandHoleShaftFit())
-
+Gui.addCommand('TechDraw_HoleShaftFit', CommandHoleShaftFit())
```
Comment (luzpaz) [inline src/Mod/TechDraw/TechDrawTools/CommandHoleShaftFit.py:67]: > Final newline missing (missing-final-newline)

Add final newline

Comment (luzpaz) [reply src/Mod/TechDraw/TechDrawTools/CommandHoleShaftFit.py:67]: I know i asked you to remove this in the first place but I didn't realize it was at the end of the file

File: src/Mod/TechDraw/TechDrawTools/CommandHoleShaftFit.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    def GetResources(self):
+        """Return a dictionary with data that will be used by the button or menu item."""
+        return {'Pixmap': 'actions/TechDraw_HoleShaftFit.svg',
+                'Accel': "",
+                'MenuText': QT_TRANSLATE_NOOP("TechDraw_HoleShaftFit", "Add hole or shaft fit"),
+                'ToolTip': QT_TRANSLATE_NOOP("TechDraw_HoleShaftFit", "Add a hole or shaft fit to a dimension<br>\
+                - select one length dimension or diameter dimension<br>\
+                - click the tool button, a panel openes<br>\
+                - select shaft fit / hole fit<br>\
+                - select the desired ISO 286 fit field using the combo box")}
```
Comment (luzpaz) [inline src/Mod/TechDraw/TechDrawTools/CommandHoleShaftFit.py:48]: Aren't we avoiding adding `<br\> in the translations ?

Comment (luzpaz) [reply src/Mod/TechDraw/TechDrawTools/CommandHoleShaftFit.py:48]: @WandererFan :point_up_2:

Comment (WandererFan) [reply src/Mod/TechDraw/TechDrawTools/CommandHoleShaftFit.py:48]: Which digit is extended in that emoji? :) 

Is it documented somewhere that tool tips shouldn't include <br>?

File: src/Mod/TechDraw/TechDrawTools/TaskHoleShaftFit.py (hunk 3 above)
Comment (benj5378) [reply src/Mod/TechDraw/TechDrawTools/TaskHoleShaftFit.py]: For some reason I didn't get a notification of your reply. Idk if it's just for me it's not intuitive, so I would probably make it a dictionary or smth like that instead, so what the variable values mean are very clear

**Comment by benj5378** [comment]
Hmm, no one ran black or flake8 :/
